pkg/sql/revert_test.go

```diff
@@ -106,6 +111,206 @@ func TestTableRollback(t *testing.T) {
+//
+// We use more splits and scatter the ranges.
+func TestRevertSpansFanout(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{UseDatabase: "test"},
+	})
+	defer tc.Stopper().Stop(context.Background())
+	s := tc.TenantOrServer(0)
+	sqlDB := tc.Conns[0]
+
+	execCfg := s.ExecutorConfig().(sql.ExecutorConfig)
+
+	db := sqlutils.MakeSQLRunner(sqlDB)
+	db.Exec(t, "CREATE DATABASE IF NOT EXISTS test")
+	db.Exec(t, "CREATE TABLE test (k INT PRIMARY KEY, rev INT DEFAULT 0, INDEX (rev))")
+
+	kvDB := s.DB()
+	desc := desctestutils.TestingGetPublicTableDescriptor(kvDB, execCfg.Codec, "test", "test")
+	span := desc.TableSpan(execCfg.Codec)
+	tableID := desc.GetID()
+
+	// Fill a table with some rows plus some revisions to those rows.
+	const (
+		initRows  = 1000
+		extraRows = 500
+		splits    = 100
+	)
+
+	db.Exec(t, "INSERT INTO test (k) SELECT generate_series(1, $1)", initRows)
+	db.Exec(t, "UPDATE test SET rev = 1 WHERE k % 3 = 0")
+	db.Exec(t, "DELETE FROM test WHERE k % 10 = 0")
+	before, ts := fingerprintTableNoHistory(t, db, tableID, "")
+	targetTime, err := hlc.ParseHLC(ts)
+	require.NoError(t, err)
+
+	beforeNumRows := db.QueryStr(t, "SELECT count(*) FROM test")
+
+	// Make some more edits: delete some rows and edit others, insert into some of
+	// the gaps made between previous rows, edit a large swath of rows and add a
+	// large swath of new rows as well.
+	db.Exec(t, "DELETE FROM test WHERE k % 5 = 2")
+	db.Exec(t, "INSERT INTO test (k, rev) SELECT generate_series(10, $1, 10), 10", initRows)
+	db.Exec(t, "INSERT INTO test (k, rev) SELECT generate_series($1+1, $1+500, 1), 500", initRows)
+	db.Exec(t, "UPDATE test SET rev = 2 WHERE k % 4 = 0")
+	db.Exec(t, "UPDATE test SET rev = 4 WHERE k > 150 and k < 350")
+
+	var splitPointsBuf strings.Builder
+	splitIncrement := (initRows + extraRows) / splits
+	for i := 0; i < 100; i++ {
+		splitPoint := splitIncrement * i
+		if i == 0 {
+			fmt.Fprintf(&splitPointsBuf, "(%d)", splitPoint)
+		} else {
+			fmt.Fprintf(&splitPointsBuf, ", (%d)", splitPoint)
+		}
+	}
+
+	// Split and scatter the table.
+	db.Exec(t, fmt.Sprintf("ALTER TABLE test SPLIT AT VALUES %s", splitPointsBuf.String()))
+	db.Exec(t, "ALTER TABLE test SCATTER")
+
+	rsCtx, close := sql.MakeJobExecContext("revert-spans", username.RootUserName(), &sql.MemoryMetrics{}, &execCfg)
+	defer close()
+
+	verifyRevert := func() {
+		reverted, _ := fingerprintTableNoHistory(t, db, tableID, "")
+		require.Equal(t, before, reverted, "expected reverted table after edits to match before")
+		db.CheckQueryResults(t, "SELECT count(*) FROM test", beforeNumRows)
+	}
+
+	t.Run("revert-fanout-with-callback", func(t *testing.T) {
+		require.NoError(t,
+			sql.RevertSpansFanout(ctx,
+				kvDB, rsCtx, []roachpb.Span{span}, targetTime, false, 10,
+				func(context.Context, roachpb.Span) error {
```

Comment:
do you think this test would add coverage if we verified in the callback that the requests were actually distributed at the kv client level?

Reply:
Always happy to add test coverage, but I'm not actually sure what you mean here. What do you have in mind?

Reply:
I'm actually not sure what i had in mind :D. This test looks fine to me.